profile/label_nodes.go

```diff
@@ -0,0 +1,97 @@
+package profile
+
+// AddLabelNodes adds psuedo-stack-frames "label:value" to each Sample with
+// labels matching the supplied keys.
+//
+// rootKeys adds frames at the root of the callgraph (first key becomes new root).
+// leafKeys adds frames at the leaf of the callgraph (last key becomes new leaf).
+//
+// Returns whether there were matches found for the label keys.
+func (p *Profile) AddLabelNodes(rootKeys, leafKeys []string) (rootm, leafm bool) {
+	// Find where to insert the new locations and functions at the end of
+	// their ID spaces.
+	var maxLocID uint64
+	var maxFunctionID uint64
+	for _, loc := range p.Location {
+		if loc.ID > maxLocID {
+			maxLocID = loc.ID
+		}
+	}
+	for _, f := range p.Function {
+		if f.ID > maxFunctionID {
+			maxFunctionID = f.ID
+		}
+	}
+	nextLocID := maxLocID + 1
+	nextFuncID := maxFunctionID + 1
+
+	// Intern the new frames we are generating.
+	locsByName := map[string]*Location{}
+	functionsByName := map[string]*Function{}
+
+	internFunction := func(functionName string) *Function {
+		function, found := functionsByName[functionName]
+		if !found {
+			function = &Function{
+				ID:   nextFuncID,
+				Name: functionName,
+			}
+			nextFuncID++
+			p.Function = append(p.Function, function)
+			functionsByName[functionName] = function
+		}
+		return function
+
+	}
+
+	internLocation := func(functionName string, function *Function) *Location {
+		location, found := locsByName[functionName]
+		if !found {
+			location = &Location{
+				ID: nextLocID,
+				Line: []Line{
+					Line{
+						Function: function,
+					},
+				},
+			}
+			nextLocID++
+			p.Location = append(p.Location, location)
+			locsByName[functionName] = location
+		}
+		return location
+	}
+
+	for _, s := range p.Sample {
+		rootsToAdd := []*Location{}
+		leafsToAdd := []*Location{}
+		for i := range rootKeys {
+			// Loop backwards, ensuring the first tag is the new root.
+			k := rootKeys[len(rootKeys)-1-i]
+			for _, v := range s.Label[k] {
```

Comment:
We seem to be added different number of frames depending on whether the tag is present and whether it has 1 or more values. Somehow, as a user, from reading the doc for the new flags I expect strictly static number of levels to be added.

Reply:
It looks like numeric labels are not supported. Note that numeric labels have units and we should support those.

Reply:
Right. In an initial draft of this CL I didn't loop over duplicate keys to make separate stack frames; instead I outputted the `strings.Join` version of them: `label1:value1,value2`. Would that be preferable? I really don't know who would use multiple values for the same key, seems like trouble.

I wonder how we should support numeric labels, I don't use them too much. Are you thinking something like `label1:1023kB`? (I hope we don't have to bucket them, that'll be complicated)

Reply:
No, not suggesting bucketing, that's too complicated. I was only thinking of `label1:1023kB`, yes. In the general case numeric labels can have high cardinality but there are important use cases where this would be useful such as the alignment label in heapz profiles.

Reply:
I've updated the CL to `strings.Join` repeated key-values.

Reply:
I've added the numeric units. This was pretty tricky to zip them together, and I think I might have found a bug in `preEncode` too, which I'll file soon.

Reply:
#651 is the bug in `preEncode`, I've added a test case that tickles the bug here.